pkg/storage/tsdb/inmemory_index_cache.go

```diff
@@ -0,0 +1,236 @@
+package tsdb
+
+import (
+	"context"
+	"reflect"
+	"unsafe"
+
+	"github.com/VictoriaMetrics/fastcache"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+
+	storecache "github.com/thanos-io/thanos/pkg/store/cache"
+	"github.com/thanos-io/thanos/pkg/tenancy"
+)
+
+type InMemoryIndexCache struct {
+	logger           log.Logger
+	cache            *fastcache.Cache
+	maxItemSizeBytes uint64
+
+	added    *prometheus.CounterVec
+	overflow *prometheus.CounterVec
+
+	commonMetrics *storecache.CommonMetrics
+}
+
+// NewInMemoryIndexCacheWithConfig creates a new thread-safe cache for index entries. It relies on the cache library
+// (fastcache) to ensures the total cache size approximately does not exceed maxBytes.
+func NewInMemoryIndexCacheWithConfig(logger log.Logger, commonMetrics *storecache.CommonMetrics, reg prometheus.Registerer, config storecache.InMemoryIndexCacheConfig) (*InMemoryIndexCache, error) {
+	if config.MaxItemSize > config.MaxSize {
+		return nil, errors.Errorf("max item size (%v) cannot be bigger than overall cache size (%v)", config.MaxItemSize, config.MaxSize)
+	}
+
+	// fastcache will panic if MaxSize <= 0.
+	if config.MaxSize <= 0 {
+		config.MaxSize = storecache.DefaultInMemoryIndexCacheConfig.MaxSize
+	}
+
+	if commonMetrics == nil {
+		commonMetrics = storecache.NewCommonMetrics(reg)
+	}
+
+	c := &InMemoryIndexCache{
+		logger:           logger,
+		maxItemSizeBytes: uint64(config.MaxItemSize),
+		commonMetrics:    commonMetrics,
+	}
+
+	c.added = promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+		Name: "thanos_store_index_cache_items_added_total",
```

Comment:
We still use `thanos_` prefix to make those metrics compatible with the previous implementation.